core/benches/bench_hashmap.rs

```diff
@@ -0,0 +1,293 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use criterion::{
+    black_box,
+    criterion_group,
+    criterion_main,
+    BatchSize,
+    Criterion,
+};
+
+use ink_core::{
+    env,
+    storage2::{
+        collections::{
+            hashmap::Entry,
+            HashMap as StorageHashMap,
+        },
+        traits::{
+            KeyPtr,
+            SpreadLayout,
+        },
+    },
+};
+use ink_primitives::Key;
+
+criterion_group!(
+    populated_cache,
+    bench_insert_populated_cache,
+    bench_remove_populated_cache,
+);
+criterion_group!(empty_cache, bench_insert, bench_remove,);
+criterion_main!(populated_cache, empty_cache,);
+
+/// The number of `ENTIRES` denotes how many test values are put into
+/// the `StorageHashMap` used in these benchmarks.
+const ENTRIES: i32 = 500;
+
+/// Returns some test values for use in benchmarks.
+fn test_values() -> Vec<(i32, i32)> {
+    (0..ENTRIES)
+        .into_iter()
+        .map(|index| (index, index))
+        .collect::<Vec<_>>()
+}
+
+/// Creates a `StorageHashMap` from the given slice.
+fn hashmap_from_slice(slice: &[(i32, i32)]) -> StorageHashMap<i32, i32> {
+    slice.iter().copied().collect::<StorageHashMap<i32, i32>>()
+}
+
+/// Creates a `StorageHashMap` from `test_values()`.
+fn setup_hashmap() -> StorageHashMap<i32, i32> {
+    let test_values = test_values();
+    hashmap_from_slice(&test_values[..])
+}
+
+/// Returns always the same `KeyPtr`.
+fn key_ptr() -> KeyPtr {
+    let root_key = Key::from([0x42; 32]);
+    KeyPtr::from(root_key)
+}
+
+/// Creates a `StorageHashMap` and pushes it to the contract storage.
+fn push_storage_hashmap() {
+    let hmap = setup_hashmap();
+    SpreadLayout::push_spread(&hmap, &mut key_ptr());
+}
+
+/// Pulls a lazily loading `StorageHashMap` instance from the contract storage.
+fn pull_storage_hashmap() -> StorageHashMap<i32, i32> {
+    <StorageHashMap<i32, i32> as SpreadLayout>::pull_spread(&mut key_ptr())
+}
+
+fn bench_insert_and_inc(hmap: &mut StorageHashMap<i32, i32>) {
+    for key in 0..ENTRIES * 2 {
+        black_box({
+            if !hmap.contains_key(&key) {
+                hmap.insert(key, key);
+            }
+            *hmap.get_mut(&key).unwrap() += 1;
+        });
```

Comment:
So I'm also still learning about `black_box`'es, but the [docs](https://docs.rs/criterion/0.2.3/criterion/fn.black_box.html) say
>[…] used to prevent the compiler from optimizing away computations in a benchmark.

In the case above we want to ensure that each operation is definitely executed. Hence it seems reasonable to me to have this scope here.